### PR TITLE
thelio-io flashing support

### PR DESCRIPTION
**Pull request by jackpot51**
Type of pull request:
- [x] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

I tested updates and downgrades on several devices.

Requires https://github.com/pop-os/system76-io-dkms/pull/3

### CODE REVIEW

**Comment by hughsie** [comment]
Also, can you squash up into one commit pls. Just force push to the branch, it's fine.

**Comment by jackpot51** [comment]
I force pushed and made the version format triplet

File: plugins/thelio-io/fu-thelio-io-device.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	if (istr == NULL)
+		return FALSE;
+
+	read_size = g_input_stream_read (istr, buf, sizeof(buf) - 1, NULL, error);
+	if (read_size < 0) {
+		fu_device_set_version (device, "0.0.0", FWUPD_VERSION_FORMAT_TRIPLET);
+	} else {
+		buf[read_size] = 0;
+		fu_device_set_version (device, (const gchar *) buf, FWUPD_VERSION_FORMAT_TRIPLET);
+	}
```
Comment (superm1) [inline plugins/thelio-io/fu-thelio-io-device.c]: Technically isn't sending buf the same in all cases? So you don't need a different case of calling set version in the < 0 case.

If it's less than zero you should be sending a buf that was initialized to zero above, no?

File: plugins/thelio-io/fu-thelio-io-device.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	/* convert GUsbDevice to GUdevDevice */
+	udev_device = fu_usb_device_find_udev_device (FU_USB_DEVICE (device), error);
+	if (udev_device == NULL)
+		return FALSE;
+
+	fn = g_build_filename (g_udev_device_get_sysfs_path (udev_device), "revision", NULL);
+	file = g_file_new_for_path (fn);
+	istr = G_INPUT_STREAM (g_file_read (file, NULL, error));
+	if (istr == NULL)
+		return FALSE;
```
Comment (superm1) [inline plugins/thelio-io/fu-thelio-io-device.c]: I'm curious, considering this is just reading in a sysfs file, why not use g_file_get_contents?

File: src/fu-usb-device.c [hunk 3]
```diff
@@ -388,6 +388,55 @@ fu_usb_device_set_dev (FuUsbDevice *device, GUsbDevice *usb_device)
...
+/**
+ * fu_usb_device_find_udev_device:
+ * @device: A #FuUsbDevice
+ * @usb_device: A #GUsbDevice, or %NULL
+ *
+ * Gets the matching #GUdevDevice for the #GUsbDevice.
+ *
+ * Returns: a #GUdevDevice, or NULL if unset or invalid
+ *
+ * Since: 1.3.1
```
Comment (superm1) [inline src/fu-usb-device.c]: 1.3.2

File: plugins/thelio-io/fu-thelio-io-device.c (hunk 2 above)
Comment (jackpot51) [reply plugins/thelio-io/fu-thelio-io-device.c]: Good point, I will use that instead

File: src/fu-usb-device.c (hunk 3 above)
Comment (jackpot51) [reply src/fu-usb-device.c]: I will change this in the next commit

Comment (hughsie) [reply src/fu-usb-device.c]: If you can push this in the next couple of days I can include this plugin in the 1.3.2 release.

Comment (jackpot51) [reply src/fu-usb-device.c]: @hughsie I have been busy with our next generation of laptops but I will try to push these changes today

**Comment by jackpot51** [comment]
@hughsie @superm1 the review comments were addressed in the last push

File: plugins/thelio-io/fu-thelio-io-device.c [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	g_autoptr(GUdevDevice) udev_device = NULL;
+
+	fu_device_add_instance_id (device, "USB\\VID_03EB&PID_2FF4");
+
+	/* convert GUsbDevice to GUdevDevice */
+	udev_device = fu_usb_device_find_udev_device (FU_USB_DEVICE (device), error);
+	if (udev_device == NULL)
+		return FALSE;
+
+	fn = g_build_filename (g_udev_device_get_sysfs_path (udev_device), "revision", NULL);
```
Comment (superm1) [inline plugins/thelio-io/fu-thelio-io-device.c]: although unlikely, you probably should check that the path you get from `g_udev_device_get_sysfs_path` is non-NULL.  (Sample code from another plugin here: https://github.com/fwupd/fwupd/blob/master/plugins/thunderbolt/fu-plugin-thunderbolt.c#L168)

File: plugins/thelio-io/fu-thelio-io-device.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	g_autoptr(FuIOChannel) io_channel = NULL;
+	g_autoptr(GUdevDevice) udev_device = NULL;
+	const guint8 buf[] = { '1', '\n' };
+
+	/* convert GUsbDevice to GUdevDevice */
+	udev_device = fu_usb_device_find_udev_device (FU_USB_DEVICE (device), error);
+	if (udev_device == NULL)
+		return FALSE;
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fn = g_build_filename (g_udev_device_get_sysfs_path (udev_device), "bootloader", NULL);
```
Comment (superm1) [inline plugins/thelio-io/fu-thelio-io-device.c]: same nit here

**Comment by superm1** [comment]
Just two very minor things I mentioned above, and I think you should add yourself to the copyright headers.

**Comment by jackpot51** [comment]
@superm1 I think I took care of those things

**Comment by jackpot51** [comment]
Fixed the build failure - g_udev_device_get_sysfs_path returns a const gchar * that does not need to be g_autofree

**Pull request merged**